app/scripts/background.js

```diff
@@ -420,15 +418,47 @@ let connectCaipMultichain;
 
 browser.runtime.onConnect.addListener(async (...args) => {
   // Queue up connection attempts here, waiting until after initialization
-  await isInitialized;
+  try {
+    await isInitialized;
 
-  // This is set in `setupController`, which is called as part of initialization
-  connectWindowPostMessage(...args);
+    // This is set in `setupController`, which is called as part of initialization
+    connectWindowPostMessage(...args);
+  } catch (error) {
+    sentry?.captureException(error);
+
+    // if have a STATE_CORRUPTION_ERROR tell the user about it and offer to
+    // restore from a backup, if we have one.
+    if (KNOWN_STATE_CORRUPTION_ERRORS.has(error.message)) {
+      const _state = await localStore.get().catch((_) => null);
+
+      const port = args[0];
+      try {
+        // send the `error` TO THE ui
+        port.postMessage({
+          data: {
+            method: METHOD_DISPLAY_STATE_CORRUPTION_ERROR,
+            params: {
+              error: {
+                message: error.message,
+                name: error.name,
+                stack: error.stack,
+              },
+              currentLocale: _state?.data?.PreferencesController?.currentLocale,
+            },
+          },
+        });
+      } catch (e) {
+        // an exception can occur here if the Window has since disconnected from
+        // the background, this is expected if the UI is closed while we are
+        // still initializing the background (during `await isInitialized;`)
+        log.error('MetaMask - Failed to send state corruption error', e);
+      }
+    }
+  }
 });
 browser.runtime.onConnectExternal.addListener(async (...args) => {
   // Queue up connection attempts here, waiting until after initialization
   await isInitialized;
```

Comment:
Question: Shall we apply the same logic for users connecting through an external dApp?

Reply:
I hadn't thought of that. Are you thinking that we should _open_ the UI to display an error when an external dApp attempts to connect, but can't because of an initialization error? Or just that we should return an error?

Reply:
I would go for "we should open the UI to display an error when an external dApp attempts to connect" to keep the behavior consistent, wdyt?

Reply:
Sounds like it'd be useful! However, I tried adding it, and it seems like dApps don't connect at all when `isInitialized` is rejected. I'm not sure why the connection between the tab and the background isn't working in this case.

I think it's fine to omit this, as the error is rare, and the user will likely click metamask to see what is happening when a dapp isn't responding.

Reply:
thanks for testing, agree we can omit it